iroh/src/protocol.rs

```diff
@@ -109,10 +113,18 @@ pub struct RouterBuilder {
 /// The protocol handler must then be registered on the node for an ALPN protocol with
 /// [`crate::protocol::RouterBuilder::accept`].
 pub trait ProtocolHandler: Send + Sync + std::fmt::Debug + 'static {
+    /// Optional interception point to handle the `Connecting` state.
+    fn on_connecting(&self, conn: Connecting) -> BoxFuture<Result<Connection>> {
```

Comment:
This is not needed anywhere. So why do we need it? For future use?

Reply:
currently this is the thing we expose, which is why I added this hook for now. don't know if it is really needed in practice

Reply:
One thing I dislike about this is that if you want to allow protocol handlers to be composed, like done in this PR, it means that you add boxing at every composition layer.

The way I usually deal with this is to have a `trait ProtocolHandler` which is not *dyn*able but has concrete return types for the futures, and then have a second `trait BoxedProtocolHandler` which has the return type fixed to BoxFuture. That way you box only once when you convert into a dyn. But that is a lot of boilerplate to write and difficult to explain to beginners.

So if we don't yet have a concrete use case for `on_connecting` I would just leave it out for now. But 🤷 ...

Reply:
> So if we don't yet have a concrete use case for `on_connecting` I would just leave it out for now. But 🤷 ...

we have now, with https://github.com/n0-computer/iroh/pull/3163